app/vmalert/group.go

```diff
@@ -277,14 +281,30 @@ var skipRandSleepOnGroupStart bool
 func (g *Group) start(ctx context.Context, nts func() []notifier.Notifier, rw *remotewrite.Client, rr datasource.QuerierBuilder) {
 	defer func() { close(g.finishedCh) }()
 
-	// Spread group rules evaluation over time in order to reduce load on VictoriaMetrics.
+	evalTS := time.Now()
+
+	// If group `eval_offset` is specified, will evaluate group
```

Comment:
I don't think eval_offset should have effect on this delay.
Let's think about this in the following perspective:
1. There is a moment of time when rule is evaluated
2. And there is a moment of time with which this rule is evaluated.

This delay here affects p1 - it spreads evaluations over time to avoid many concurrent requests. 
But p2 is what actually eval_offset should have effect on. If I have a rule with `interval: 1h` and I want it to be evaluated on the 5th minute of the hour - I'm going to set `eval_offset: 5m`. But, this has nothing to do with p1. This rule can be evaluated on any moment of time on interval between 5th and 60th min of the current hour. I can evaluate it at 11:49 with timestamp 11:05, and it still will be correct.

So what I propose is to propagate eval_offect as one of https://github.com/VictoriaMetrics/VictoriaMetrics/blob/0df506de54df5c24ef040fc2363ff402d5e80eb3/app/vmalert/datasource/vm.go#L33 new params. This new param will be checked here https://github.com/VictoriaMetrics/VictoriaMetrics/blob/0df506de54df5c24ef040fc2363ff402d5e80eb3/app/vmalert/datasource/vm_prom_api.go#L167-L169  and if it is not nil - the current timestamp should be truncated by eval_offset.

It is also interesting that `datasource.queryTimeAlignment` kinda conflicts with eval_offset. I propose to resolve the conflict in the following way. At https://github.com/VictoriaMetrics/VictoriaMetrics/blob/0df506de54df5c24ef040fc2363ff402d5e80eb3/app/vmalert/datasource/vm_prom_api.go#L167-L169  we'll do the following check:
1. If eval_offset is empty and `datasource.queryTimeAlignment=true` - we truncate timestamp with eval_interval
2.  If eval_offset is empty and `datasource.queryTimeAlignment=false` - we do nothing
3. If eval_offset is **not empty** and `datasource.queryTimeAlignment=true`- we truncate timestamp with eval_offset
4. If eval_offset is **not empty** and `datasource.queryTimeAlignment=false`- we truncate timestamp with eval_offset

wdyt @Haleygo ?

Reply:
>This rule can be evaluated on any moment of time on interval between 5th and 60th min of the current hour. I can evaluate it at 11:49 with timestamp 11:05, and it still will be correct.

Yeah, we can have rule evaluated at 11:49 with timestamp 11:05, but the result will only show up after 11:49 in that case, so it's still uncontrollable from my perspective. I use this `eval_offset` as a force offset for group evaluation in range of [0...group_interval] after aligned with interval.

>It is also interesting that datasource.queryTimeAlignment kinda conflicts with eval_offset. 
I mixed them before, then I divided their impacts, `eval_offset` only for evaluation time, `queryTimeAlignment` only for raw series timestamp of evaluation. 
If `eval_offset` is setting, the evaluation result is always consistent and predictable without effects from different vmalert or start time, so `queryTimeAlignment` is not needed. But I didn't see much influences here and left it as before.

Reply:
I think let `eval_offset` to control the p1 is somehow more clearly for users, like the original issue asked.
>If we could give it an evaluation-offset of 2m or something like that, that would make sure it evaluates and writes the result within the first 5m of the hour

People would want to see results at 11:02 when they set 2m offset instead of having unknown delay on evaluation.

Reply:
> but the result will only show up after 11:49 in that case

just to clarify: yes, it will be queryable after 11:49, but it will have 11:05 timestamp.

Although, this 11:49 thing is what I'd like to fix afterwards. I think we should have a 5-10 min cap on the delay. For example, if we have 10 rules with interval 1h - all of them should start in first, let's say, 5min after vmalert start. I believe 5min interval is enough period to spread the load. So once we have this cap, the argument about delay becomes less relevant.

Reply:
> People would want to see results at 11:02 when they set 2m offset instead of having unknown delay on evaluation.

I don't think people care that much about waiting for each result to appear. They care only about right timestamps and calculation on the chart.

Reply:
>All rules for a group with eval_offsset will start in the same moment of time

Actually rules in a group are evaluated at same time anyway. But mulitple groups with same offset will be crowded at same time if misconfigured.
But I get your points now, maybe my group example with 1h interval is too big, and took the deviation between 11:49 and 11:05 too serious.

Reply:
> Actually rules in a group are evaluated at same time anyway. But mulitple groups with same offset will be crowded at same time if misconfigured.

yeah, my bad. Thanks for pointing that out

Reply:
>then we'll have the following consequences:

All rules for a group with eval_offsset will start in the same moment of time
There will be a conflict between the moment when group was evaluated and -datasource.queryTimeAlignment param
There will be a conflict between the moment when group was evaluated and -datasource.lookback param
Rules executed at timestamps close to the realtime are more likely to get a partial result, as data usually appears in the datasource with delay.

For now, this pr will address those problems:
1. All the group with same `eval_offset`  won’t be start at the same moment when vmalert restart, there will still be randSleep on each of group to spread the pressure like before.
2. Both `eval_offset` and `-datasource.queryTimeAlignment` have the power to influence the result of group evaluation.  `eval_offset`  will do it on group level, if `eval_offset`  is used, group’s evaluation results will always be at the exact offset in the range of [0...interval].
   `-datasource.queryTimeAlignment` will align the evaluation timestamp globally like before, but has no influence on group with `eval_offset`
3. `-datasource.lookback` won’t work when `eval_offset` is used. If users see partial results on group with `eval_offset`, it could be a rather big delay on the original data in datasource, it’s recommended to adding `-search.latencyOffset` on group’s params instead of using `lookback` , cause `lookback` will mess up the timestamp that result should have.
   FYI, this situation should be rarely, since group with `eval_offset` will sleep enough time to make sure the evaluation request happens after the offset ts. But if users do have a big delay on their original data, for now, they will have to set `search.latencyOffset`  for groups with `eval_offset` and `-datasource.lookback` for groups without `eval_offset`
4. see p3